File: archive/cpp/array/LargestRectangleInHistogram.cpp
class Solution {
public:
    int largestRectangleArea(std::vector<int> &height) {
        height.push_back(0);
        int i = 0;
        int result = 0;
        std::stack<int> stack;
        while (i < int(height.size())) {
            if (stack.empty() || height[stack.top()] <= height[i]) {
                stack.push(i);
                i ++;
            } else {
                int length = height[stack.top()];
                stack.pop();
                int width;
                if (stack.empty()) {
                    width = i;
                } else {
                    width = i - stack.top() - 1;
                }
                result = std::max(length * width, result);
            }
        }
        return result;
    }
};

File: archive/python/array/LargestRectangleInHistogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        height.append(0)
        result = 0
        stack = []
        i = 0
        while i < len(height):
            if len(stack) == 0 or height[stack[-1]] <= height[i]:
                stack.append(i)
                i += 1
            else:
                length = height[stack[-1]]
                stack.pop()
                if len(stack) > 0:
                    width = i - stack[-1] - 1
                else:
                    width = i
                result = max(length * width, result)
        return result

if __name__ == "__main__":
    solution = Solution()

    print solution.largestRectangleArea([1])
